test/parallel-benchmark/mzcompose.py

```diff
@@ -0,0 +1,1186 @@
+            cur.execute(self.query)
+        conn.close()
+
+    def __str__(self) -> str:
+        return f"{self.query} (standalone)"
+
+
+class ReuseConnQuery(Action):
+    def __init__(
+        self,
+        query: str,
+        conn_info: PgConnInfo,
+        strict_serializable: bool = True,
+    ):
+        self.query = query
+        self.conn = conn_info.connect()
+        self.conn.autocommit = True
+        self.cur = self.conn.cursor()
+        if not strict_serializable:
+            self.cur.execute("SET TRANSACTION_ISOLATION TO 'SERIALIZABLE'")
+
+    def _run(self, conns: queue.Queue):
+        self.cur.execute(self.query)
+
+    def __str__(self) -> str:
+        return f"{self.query} (reuse connection)"
+
+
+class PooledQuery(Action):
+    def __init__(self, query: str, conn_info: PgConnInfo):
+        self.query = query
+        self.conn_info = conn_info
+
+    def _run(self, conns: queue.Queue):
+        conn = conns.get()
+        with conn.cursor() as cur:
+            cur.execute(self.query)
+        conns.task_done()
+        conns.put(conn)
+
+    def __str__(self) -> str:
+        return f"{self.query} (pooled)"
+
+
+class Distribution:
+    def generate(self, duration: int) -> Iterator[None]:
+        raise NotImplementedError
+
+
+class Periodic(Distribution):
+    """Run the action in each thread in one second, spread apart by the 1/per_second"""
+
+    def __init__(self, per_second: int):
+        self.per_second = per_second
+
+    def generate(self, duration: int) -> Iterator[None]:
+        sleep_time = 1 / self.per_second
+        for i in range(duration * self.per_second):
+            yield
+            time.sleep(sleep_time)
+
+
+class Gaussian(Distribution):
+    """Run the action with a sleep time between actions drawn from a Gaussian distribution"""
+
+    def __init__(self, mean: float, stddev: float):
+        self.mean = mean
+        self.stddev = stddev
+
+    def generate(self, duration: int) -> Iterator[None]:
+        end_time = datetime.datetime.now() + datetime.timedelta(seconds=duration)
+        while datetime.datetime.now() < end_time:
+            yield
+            sleep_time = max(0, random.gauss(self.mean, self.stddev))
+            time.sleep(sleep_time)
+
+
+class PhaseAction:
+    def run(
+        self,
+        duration: int,
+        jobs: queue.Queue,
+        conns: queue.Queue,
+        measurements: collections.defaultdict[str, list[Measurement]],
+    ) -> None:
+        raise NotImplementedError
+
+
+class OpenLoop(PhaseAction):
+    def __init__(self, action: Action, dist: Distribution):
+        self.action = action
+        self.dist = dist
+
+    def run(
+        self,
+        duration: int,
+        jobs: queue.Queue,
+        conns: queue.Queue,
+        measurements: collections.defaultdict[str, list[Measurement]],
+    ) -> None:
+        for _ in self.dist.generate(duration):
+            jobs.put(lambda: self.action.run(conns, measurements))
```

Comment:
This might not measure the desired latency. In open-loop scenarios, what we want is the time from request generation to fulfillment, but here we seem to measure the time from local dequeue until fulfillment, which can be considerably lower if the requests spends time in the local queue.